controllers/volumereplicationgroup_controller.go

```diff
@@ -469,34 +466,41 @@ func (v *VRGInstance) createVolumeReplicationResources() bool {
 		volRep := &volrep.VolumeReplication{}
 
 		err := v.reconciler.Get(v.ctx, pvcNamespacedName, volRep)
-		if err != nil && errors.IsNotFound(err) {
-			log.Info("Uploading PersistentVolume")
-
-			if err = v.uploadPV(pvc); err != nil {
-				log.Error(err, "failed to upload PV metadata")
-
+		if err != nil {
+			if errors.IsNotFound(err) {
+				requeue = v.createVolumeReplication(pvc)
+			} else {
+				// requeue on failure to ensure any PVC not having a corresponding VR CR
 				requeue = true
-
-				// There's no point enabling PV data replication if
-				// replication of PV metadata to object store has failed.
-				break
+				log.Error(err, "failed to get VolumeReplication resource")
 			}
+		}
+	}
 
-			log.Info("Creating VolumeReplication resource for PersistentVolumeClaim")
+	return requeue
+}
 
-			if err = v.createVolumeReplicationForPVC(pvc); err == nil {
-				continue
-			}
-		}
+func (v *VRGInstance) createVolumeReplication(
+	pvc corev1.PersistentVolumeClaim) (requeue bool) {
+	requeue = false
 
-		log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
-			"errorValue", err)
+	// Prior to creating VR, upload PV metadata to object store
+	if err := v.uploadPV(pvc); err != nil {
```

Comment:
I flipped replication first and then upload, as an uploaded PV can be made available in the peer on a failover, without the storage having been replicated.

In the current situation, if storage is replicated, but PV not yet backed up failover would create a new PV at the other end, and the intention is that this PVC was never protected.

We do not need to change it in this PR, and will be handled in PR #37

Reply:
I understand that you're speaking about the order of PV upload and enabling PV replication, but I'm not exactly sure what you're proposing.  I believe we should first upload the PV and if the upload fails, we should not enable replication.  I believe this is what I've coded.  At the same time, I don't think you're planning to change the order to first enable replication and then uploadPV?

Reply:
I am proposing replication first and then PV upload:
- Say we upload the PV and then fail to enable replication, then on failover (assuming we restore the PV) the failover site PVC that binds to this PV will fail to come up
- If we do the other way around, and fail to upload the PV, there will be a new PV created at the target for the PVC and we lose any data that was written to the older primary

The end solution may lie in choosing which PVs to restore, if we use status of VRG at AVR and leverage the same we may ensure both cases are clean.